### PR TITLE
[ROCm][Perf][GLM-5.3-Flash] Add SparseIndexerTopk AITER backend

### LINKED ISSUE #58170: [ROCm][Perf]: GLM-5.3-Flash should use AITER topk kernel during decode
### 🚀 The feature, motivation and pitch

This could improve performance

### Alternatives

_No response_

### Additional context

_No response_

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by simondanielsson**
## Purpose

GLM-5.3-Flash indexer dispatches to `torch.ops._C.top_k_per_row_decode` during decode. This leads to two calls to HIP kernel `topKPerRowDecode512DeviceLengthAware` which is quite slow for this model. 

We change it to an AITER kernel instead by adding a new backend to the SparseIndexerTopk op and dispatching to it through `get_indexer_topk(topk_backend)` for GLM-5.3-Flash. 

Models like DSv4 keep their old path.

### 2. GLM-5.3-Flash perf diff on vllm bench serve sweep (MI350)

```
VLLM_ROCM_USE_AITER=1 vllm serve zai-org/GLM-5.3-Flash \
  --tensor-parallel-size 8 \
  --max-num-seqs 512 \
  --attention-backend ROCM_AITER_MLA_SPARSE  
  --sparse-indexer-topk-backend [auto|per_row]
```

### 1. GSM8k

This branch

Main
|Tasks|Version|     Filter     |n-shot|  Metric   |   |Value |   |Stderr|
|-----|------:|----------------|-----:|-----------|---|-----:|---|-----:|
|gsm8k|      3|flexible-extract|     5|exact_match|↑  |0.9219|±  |0.0074|
|     |       |strict-match    |     5|exact_match|↑  |0.9196|±  |0.0075|

### 2. GLM-5.3-Flash perf diff on vllm bench serve sweep (MI350)

todo

### 3. Non-regression of DSv4-Flash vs main, bench serve sweep

todo

Fixes https://github.com/vllm-project/vllm/issues/58170.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [ ] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
</details>

### CODE REVIEW

File: vllm/_aiter_ops.py [hunk 1]
```diff
@@ -2472,6 +2477,98 @@ def fused_moe_supports_heterogeneous_shared_expert(cls, num_tokens: int) -> bool
...
+        """Whether AITER's sparse indexer top-k beats the in-tree kernel for
+        this shape. The in-tree decode kernel's advantage window was measured
+        on DSV4's compressed-KV logits, so it applies to ``indexer="dsa"``
+        only.
+        """
+        if compress_ratio <= 1 or not cls.is_indexer_top_k_enabled():
+            return False
+
+        if not is_prefill:
+            assert max_valid_seq_len is not None
```
Comment (simondanielsson) [inline vllm/_aiter_ops.py]: **Note**: we preserve this heuristic by default (it was tuned for dsv4) by setting indexer="dsa" inside rocm_aiter_sparse_mla.py

File: vllm/v1/attention/ops/rocm_aiter_mla_sparse.py [hunk 2]
```diff
@@ -71,43 +56,6 @@ def _indexer_k_is_c4a_block_flat(compress_ratio: int) -> bool:
...
-    num_columns: int | None = None,
-    topk_tokens: int = 1024,
-    on_gfx950: bool = _ON_GFX950,
-) -> Callable[..., None] | None:
-    if compress_ratio <= 1 or not on_gfx950:
-        return None
-
-    if not is_prefill:
-        assert max_valid_seq_len is not None
-        if (
```
Comment (simondanielsson) [inline vllm/v1/attention/ops/rocm_aiter_mla_sparse.py:89]: **Note**: moved this dsv4-specific heuristic into `dsv4_indexer_prefers_native_top_k` instead

File: vllm/model_executor/layers/indexer_topk.py [hunk 3]
```diff
@@ -193,6 +199,15 @@ def resolve_backend(
                     f"inputs violate DeepSelect's constraints: dtype={logits.dtype},"
                     f" stride={logits.stride()}, topk={topk_tokens}"
                 )
+        elif self._backend == "aiter":
+            if not self._is_rocm:
+                failures.append("requires a ROCm platform")
+            elif not self._aiter_capable:
+                failures.append("requires AITER enabled on gfx950")
```
Comment (dllehr-amd) [inline vllm/model_executor/layers/indexer_topk.py]: Right now we're making an assumption that not being on gfx950 was the reason for _aiter_capable was false.  While technically this is true as of now, any additional support for the topk indexer will need to have this code touched as well

**Comment by dllehr-amd** [comment]
As a future TODO would it be worth creating a backend that can decide to use the aiter side, or the `_per_row` at runtime?

**Comment by dllehr-amd** [comment]
@simondanielsson  for larger sizes can we update the recipes project to enable this?

File: vllm/models/glm5next/amd/sparse_indexer.py [hunk 4]
```diff
@@ -87,6 +92,41 @@ def _kpool_compress_insert(
...
+        compress_ratio=index_kpool,
+        num_rows=num_rows,
+        max_valid_seq_len=max_valid_seq_len,
+    ):
+        return "aiter"
+    if (
+        index_kpool > 1
+        and select_k == 512
+        and _GFX950_KPOOL_AITER_MIN_POOLS
+        <= max_valid_seq_len
```
Comment (simondanielsson) [inline vllm/models/glm5next/amd/sparse_indexer.py:133]: I need to check this does what I expect it to do

File: vllm/models/glm5next/amd/sparse_indexer.py [hunk 5]
```diff
@@ -580,15 +621,28 @@ def sparse_attn_indexer_kpool(
         else:
             topk_dst = topk_indices_buffer[:num_padded_tokens, :topk_tokens]
 
-        torch.ops._C.top_k_per_row_decode(
+        # FULL graphs are not keyed by context length, so the kernel choice
+        # baked in at capture must hold for any replay: use the worst-case
+        # bound instead of this batch's context length.
+        if get_forward_context().cudagraph_runtime_mode == CUDAGraphMode.FULL:
```
Comment (dllehr-amd) [inline vllm/models/glm5next/amd/sparse_indexer.py]: can you review this area again real quick?

**Comment by simondanielsson** [comment]
@dllehr-amd 

1. Updated recipe: https://github.com/vllm-project/recipes/pull/1025
2. Now FULL cudagraphs _never_ dispatch to the AITER backend as it cannot discern between long/short ISL's, so it requires manual opt in via the CLI. The heuristic "enabled if >=64k ISL" is still on though if you would for some reason run decode without FULL cudagraphs. See commit https://github.com/vllm-project/vllm/pull/58167/changes/551a9e1174935dd8b16023ff15e634fc02687319

PTAL 🙏

File: vllm/models/glm5next/amd/sparse_indexer.py (hunk 5 above)
Comment (simondanielsson) [reply vllm/models/glm5next/amd/sparse_indexer.py]: Updated: see comment below https://github.com/vllm-project/vllm/pull/58167#issuecomment-5801608894
